File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/PointerInputState.java
package io.appium.espressoserver.lib.helpers.w3c.state;

import java.util.HashSet;
import java.util.Set;

import io.appium.espressoserver.lib.helpers.w3c.models.InputSource.PointerType;

/**
 * Pointer input state specified in 17.2 of spec
 *
 * https://www.w3.org/TR/webdriver/#input-source-state
 */
public class PointerInputState implements InputState {
    private Set<Integer> pressed = new HashSet<>();
    private PointerType type;
    private long x = 0;
    private long y = 0;

    public PointerInputState(PointerType pointerType) {
        this.type = pointerType;
    }

    public boolean isPressed(int num) {
        return pressed.contains(num);
    }

    public void addPressed(int num) {
        pressed.add(num);
    }

    public void removePressed(int num) {
        pressed.remove(num);
    }

    public Set<Integer> getButtons() {
        return pressed;
    }

    public boolean hasPressedButtons() {
        return !pressed.isEmpty();
    }

    public long getX() {
        return x;
    }

    public void setX(long x) {
        this.x = x;
    }

    public long getY() {
        return y;
    }

    public void setY(long y) {
        this.y = y;
    }

    public PointerType getType() {
        if (type == null) {
            return PointerType.TOUCH;
        }
        return type;
    }

    public void setType(PointerType type) {
        this.type = type;
    }

    public String logMessage() {
        return String.format(
                "pointer-type=[%s] x=[%s] y=[%s] pressed=[%s]",
                type, x, y, pressed
        );
    }
}

File: espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/helpers/w3c/state/PointerInputState.kt
package io.appium.espressoserver.lib.helpers.w3c.state

import io.appium.espressoserver.lib.helpers.w3c.models.InputSource
import java.util.*

/**
 * Pointer input state specified in 17.2 of spec
 *
 * https://www.w3.org/TR/webdriver/#input-source-state
 */
class PointerInputState : InputState {
    private val pressed: MutableSet<Int> = HashSet()
    var x = 0f
    var y = 0f
    private var _type: InputSource.PointerType? = null
    @Suppress("SuspiciousVarProperty")
    var type: InputSource.PointerType? = null
        get() { return _type ?: InputSource.PointerType.TOUCH }

    @Suppress("ConvertSecondaryConstructorToPrimary")
    constructor(type: InputSource.PointerType?) {
        this._type = type
    }

    fun isPressed(num: Int): Boolean {
        return pressed.contains(num)
    }

    fun addPressed(num: Int) {
        pressed.add(num)
    }

    fun removePressed(num: Int) {
        pressed.remove(num)
    }

    val buttons: Set<Int>
        get() = pressed

    fun hasPressedButtons(): Boolean {
        return pressed.isNotEmpty()
    }

    fun logMessage(): String {
        return String.format(
                "pointer-type=[%s] x=[%s] y=[%s] pressed=[%s]",
                type, x, y, pressed
        )
    }

}